### PR TITLE
add federated learning parameter server(fl-ps) mode

### PR DESCRIPTION
**Pull request by ziyoujiyi**
### PR types
New features
### PR changes
Others
### Describe
1. add federated learning model 
2. add fl-ps train.py
3. add fl-ps split program pass
4. add fl-ps unittest
5. support N cpus + N cpus
6. test ok

### CODE REVIEW

**Review by fuyinno4 — DISMISSED** [review]

**Review by fuyinno4 — APPROVED** [review]

File: paddle/fluid/framework/data_feed.cc [hunk 1]
```diff
@@ -971,18 +975,33 @@ void MultiSlotDataFeed::PutToFeedVec(
...
     const auto& type = ins_vec[i].GetType();
     const auto& offset = ins_vec[i].GetOffset();
     int total_instance = static_cast<int>(offset.back());
-
+    VLOG(4) << "total_instance: " << total_instance;
+    // platform::CPUPlace()
+    VLOG(4) << "this->place_: " << this->place_;
     if (type[0] == 'f') {  // float
       const auto& feasign = ins_vec[i].GetFloatData();
+      /*
```
Comment (fuyinno4) [inline paddle/fluid/framework/data_feed.cc]: remove

File: python/paddle/distributed/ps/the_one_ps.py [hunk 2]
```diff
@@ -1010,19 +1015,23 @@ def sync_strategy_envs():
...
                         self._init_all_params(scopes, send_ctx, dense_map)
 
-            fleet.util.barrier()
-        if not self.context['use_ps_gpu']:
+            fleet.util.barrier()  # 保证 0 号 worker 参数 push_dense_param over
+
+        if self.is_heter_ps_mode == False or not self.context['use_ps_gpu']:
             self._pull_all_dense(scopes, send_ctx, dense_map)
-        fleet.util.barrier()
+            fleet.util.barrier()
```
Comment (fuyinno4) [inline python/paddle/distributed/ps/the_one_ps.py]: check this with guanqun?

File: python/paddle/distributed/ps/utils/ps_program_builder.py [hunk 3]
```diff
@@ -90,7 +94,6 @@ def _build_programs(self):
 
 class GeoPsProgramBuilder(PsProgramBuilder):  # 仅 CPU 模式
     def __init__(self, pass_ctx):
-        logger.info("start building geo-ps program")
```
Comment (fuyinno4) [inline python/paddle/distributed/ps/utils/ps_program_builder.py:93]: check with caibei ?

**Review by fuyinno4 — CHANGES_REQUESTED** [review]

File: paddle/fluid/framework/data_feed.cc (hunk 1 above)
Comment (ziyoujiyi) [reply paddle/fluid/framework/data_feed.cc]: deleted

**Review by zmxdream — APPROVED** [review]

File: python/paddle/fluid/executor.py [hunk 4]
```diff
@@ -2341,20 +2349,6 @@ def start_heter_trainer(self,
...
                             fetch_handler=None):
-        return self._start_heter_trainer(program, scope, False, debug,
-                                         fetch_list, fetch_info, print_period,
-                                         fetch_handler)
-
-    def _start_heter_trainer(self,
-                             program=None,
-                             scope=None,
-                             is_infer=False,
-                             debug=False,
```
Comment (zmxdream) [inline python/paddle/fluid/executor.py:2352]: 这块找xinxuan确认下

File: paddle/fluid/framework/heter_pipeline_trainer.cc [hunk 5]
```diff
@@ -308,5 +333,5 @@ Scope* HeterPipelineTrainer::GetWorkerScope(int thread_id) {
 }
 
 }  // end namespace framework
-}  // end namespace paddle
+}  // namespace paddle
```
Comment (zmxdream) [inline paddle/fluid/framework/heter_pipeline_trainer.cc:336]: 不应该去掉

File: python/paddle/fluid/executor.py (hunk 4 above)
Comment (ziyoujiyi) [reply python/paddle/fluid/executor.py:2352]: done

File: paddle/fluid/framework/heter_pipeline_trainer.cc (hunk 5 above)
Comment (ziyoujiyi) [reply paddle/fluid/framework/heter_pipeline_trainer.cc:336]: done

File: python/paddle/fluid/executor.py [hunk 6]
```diff
@@ -2365,7 +2359,7 @@ def _start_heter_trainer(self,
             fetch_info=fetch_info,
             print_period=print_period)
 
-        trainer._set_infer(is_infer)
+        trainer._set_infer(False)
```
Comment (fuyinno4) [inline python/paddle/fluid/executor.py:2362]: hardcode？

**Review by fuyinno4 — APPROVED** [review]

File: python/paddle/distributed/ps/the_one_ps.py [hunk 7]
```diff
@@ -1010,19 +1015,27 @@ def sync_strategy_envs():
...
                 if not self.context['use_ps_gpu']:
                     if role_id == 0:
+                        print("entering self._init_all_params()")
                         self._init_all_params(scopes, send_ctx, dense_map)
 
-            fleet.util.barrier()
+            fleet.util.barrier()  # 保证 0 号 worker 参数 push_dense_param over
+
         if not self.context['use_ps_gpu']:
-            self._pull_all_dense(scopes, send_ctx, dense_map)
```
Comment (zhaocaibei123) [inline python/paddle/distributed/ps/the_one_ps.py:1022]: if xx mode:
    push dense param
    barrier
    pull dense
elif xx1 mode:
    push dense param1
    barrier
    pull dense1
else ...

File: python/paddle/distributed/ps/utils/ps_program_builder.py (hunk 3 above)
Comment (ziyoujiyi) [reply python/paddle/distributed/ps/utils/ps_program_builder.py:93]: done

File: python/paddle/fluid/tests/unittests/ps/dataset_generator_A.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/fluid/tests/unittests/ps/dataset_generator_A.py:1]: Copyright (c) 2022

File: python/paddle/fluid/tests/unittests/ps/dataset_generator_B.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/fluid/tests/unittests/ps/dataset_generator_B.py:1]: Copyright (c) 2022

File: python/paddle/fluid/tests/unittests/ps/test_fl_ps.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Ligoml) [inline python/paddle/fluid/tests/unittests/ps/test_fl_ps.py:3]: Copyright (c) 2022

**Review by Ligoml — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**